src/v/cluster/tm_stm.cc

```diff
@@ -68,42 +68,61 @@ std::optional<tm_transaction> tm_stm::get_tx(kafka::transactional_id tx_id) {
     return std::nullopt;
 }
 
-ss::future<bool> tm_stm::barrier() {
+ss::future<checked<model::term_id, tm_stm::op_status>> tm_stm::barrier() {
+    if (!_c->is_leader()) {
+        return ss::make_ready_future<
+          checked<model::term_id, tm_stm::op_status>>(
+          tm_stm::op_status::not_leader);
+    }
     if (_insync_term != _c->term()) {
-        return ss::make_ready_future<bool>(false);
+        return ss::make_ready_future<
+          checked<model::term_id, tm_stm::op_status>>(
+          tm_stm::op_status::not_leader);
     }
-    auto term = _c->term();
+    auto term = _insync_term;
     return quorum_write_empty_batch(model::timeout_clock::now() + _sync_timeout)
       .then_wrapped([this, term](ss::future<result<raft::replicate_result>> f) {
           try {
               if (!f.get0().has_value()) {
-                  return false;
+                  return ss::make_ready_future<
+                    checked<model::term_id, tm_stm::op_status>>(
+                    tm_stm::op_status::unknown);
               }
               if (term != _c->term()) {
-                  return false;
+                  return ss::make_ready_future<
+                    checked<model::term_id, tm_stm::op_status>>(
+                    tm_stm::op_status::unknown);
               }
-              return true;
+              return ss::make_ready_future<
+                checked<model::term_id, tm_stm::op_status>>(term);
           } catch (...) {
               vlog(
                 clusterlog.error,
                 "Error during writing a barrier batch: {}",
                 std::current_exception());
-              return false;
+              return ss::make_ready_future<
+                checked<model::term_id, tm_stm::op_status>>(
+                tm_stm::op_status::unknown);
```

Comment:
does `return tm_stm::op_status::unknown` not auto convert into the ready future (like it seems to below in `sync` with `co_return`?

Reply:
done